db/db.ts

```diff
@@ -525,6 +527,45 @@ export async function getFlatTagGraph(knex: KnexReadonlyTransaction): Promise<
     return { ...tagGraphByParentId, __rootId: tagGraphRootIdResult.id }
 }
 
+// DFS through the tag graph and create a map of parent tags for each child tag
+// e.g. { "Child": [ "Parent", "Grandparent" ], "Parent": [ "Grandparent" ] }
+// parent tags are listed in no particular order
+export async function getParentTagsByChildName(
+    trx: KnexReadonlyTransaction
+): Promise<Record<DbPlainTag["name"], DbPlainTag["name"][]>> {
+    const { __rootId, ...flatTagGraph } = await getFlatTagGraph(trx)
+    const tagGraph = createTagGraph(flatTagGraph, __rootId)
+
+    const tagsById = await trx("tags")
+        .select("id", "name")
+        .then((tags) => keyBy(tags, "id"))
+
+    const parentTagsByChildName: Record<
+        DbPlainTag["name"],
+        DbPlainTag["name"][]
+    > = {}
+
+    function trackParents(node: TagGraphNode): void {
+        for (const child of node.children) {
+            trackParents(child)
+        }
+
+        const preexistingParents = parentTagsByChildName[node.name] ?? []
+        // node.path is an array of tag ids from the root to the current node
+        // slice to remove the root node and the current node, then map them into tag names
+        const newParents = node.path.slice(1, -1).map((id) => tagsById[id].name)
+
+        parentTagsByChildName[node.name] = uniq([
+            ...preexistingParents,
+            ...newParents,
+        ])
+    }
+
+    trackParents(tagGraph)
+
+    return parentTagsByChildName
+}
+
```

Comment:
I would let the database do this, and replace it with one sequel query. MySQL 8 supports recursive common table expressions (the thing in the `WITH` block). It looks like this:

```sql
WITH RECURSIVE flat_tags AS (
  -- base case
  SELECT * FROM tags WHERE slug in ('population-growth', 'urbanization')

  UNION ALL

  -- recursive case
  SELECT t.*
  FROM tags t
  INNER JOIN flat_tags ft ON (t.id = ft.parentId)
)
SELECT *
FROM flat_tags
```

Reply:
Here's the query that approxmately creates the same structure as this TS function, though it also includes tags that aren't children of the tag-graph-root.

```sql
WITH RECURSIVE ancestors AS (
SELECT
	tg.childId,
	tg.parentId,
	t_parent.name AS parent_name,
	t_child.name AS child_name
FROM
	tag_graph tg
JOIN tags t_parent ON
	tg.parentId = t_parent.id
JOIN tags t_child ON
	tg.childId = t_child.id
WHERE
	t_parent.name != "tag-graph-root"
UNION
SELECT
	a.childId,
	a.parentId,
	t_parent.name AS parent_name,
	a.child_name
FROM
	ancestors a
JOIN tag_graph tg ON
	a.parentId = tg.childId
JOIN tags t_parent ON
	tg.parentId = t_parent.id
WHERE
	t_parent.name != "tag-graph-root"
)
SELECT
	a.child_name,
	GROUP_CONCAT(a.parent_name ORDER BY a.parentId ASC SEPARATOR ', ') AS ancestors
FROM
	ancestors a
GROUP BY
	a.child_name;
```

idk if this is any easier to work with, and I don't think performance is an issue. Do you strongly feel like we should be doing this in SQL?